source/simulator/introspection.cc

```diff
@@ -399,10 +405,7 @@ namespace aspect
   bool
   Introspection<dim>::composition_type_exists (const CompositionalFieldDescription::Type &type) const
   {
-    for (unsigned int c=0; c<composition_descriptions.size(); ++c)
-      if (composition_descriptions[c].type == type)
-        return true;
-    return false;
+    return composition_indices_for_type[type].size() > 0;
```

Comment:
Just add an assert here that `type` is a valid index into the array. This way we'll catch the case where we add an `enum` element but forget to update `n_types`. Same in the other places where you use `type` as an index.